[Aw Joey] iP

AwJoey opened the pull request:
# Axolotl
>  “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Axolotl your mind of having to remember things you need to do. It's,
- text-based
* easy to learn
+ ~~FAST~~ *SUPER FAST* to use

All you need to do is,
1. download it from [here](https://github.com/AwJoey/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the ```main``` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

AwJoey committed:
edited .gitignore file to ignore .class files

AwJoey committed:
chatbot renamed to axolotl and added greet and exit messages

AwJoey committed:
edited design of greeting messages

AwJoey committed:
optimizing input and print speed

AwJoey committed:
echo commands entered by user

AwJoey committed:
store user inputs into list and list can be displayed

AwJoey committed:
Created Task class as tasks have statuses now

AwJoey committed:
Added the ability to mark tasks as done/undone

AwJoey committed:
Made changes to the method of task status retrieval so as to reduce repetition

AwJoey committed:
Implemented new types of Tasks (ToDo, Deadline, Event) and edited chatbot's commands to cater to them

AwJoey committed:
set up automated text UI testing

AwJoey committed:
handled some errors that may occur and customized error messages for each case

AwJoey committed:
discovered more errors and handled them

AwJoey committed:
Implemented delete function to chatbot

AwJoey committed:
Implemented enumerators for command types for easy additiona nd modification in future

AwJoey committed:
Revert "Implemented enumerators for command types for easy additiona nd modification in future"

This reverts commit 61a0a84d0384ff183609373a9d3a4a58cb919c67.

AwJoey committed:
Reapply "Implemented enumerators for command types for easy additiona nd modification in future"

This reverts commit 385595d0df1d9034a8405eed0ce71246cd167fe0.

AwJoey committed:
Add DukeExcpetion.java

To handle general exceptions in the application.

AwJoey committed:
Add Storage.java: to consolidate IO file functions

Make code easier to find and implement for future edits and debug.

AwJoey committed:
Edit Event.java, Task.java, ToDo.java: added toStore() method

Task subclasses have some common behaviors.

The common behaviors across different Task subtypes cause code
duplication.

Extracting the common behavior into a super class allows us to use
polymorphism when storing file content.

Let's pull up storage behaviors common to all task subclasses into a new
parent class named Task.

Using inheritance is preferable over composition in this situation
because the common behaviors are not composable.

AwJoey committed:
Edit Duke.java: Implement save tasks feature

AwJoey committed:
Edit Event.java, Task.java, Deadline.java: edited dates from String to
DateTime

AwJoey committed:
Edit Duke.java, Storage.java: edited methods involving addition of Tasks
which involve time (Deadline, Event)

AwJoey committed:
Edit Duke.java: Caught IndexOutOfBounds for Event entry format

jinhanfromNUS commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,193 @@
+import java.util.ArrayList;
+import  java.util.EnumSet;
+
+enum Command {
```
I like how you create this enum Command, it makes the code very readable.

jinhanfromNUS commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,193 @@
+        this.commandSignature = commandSignature;
+    }
+}
 public class Duke {
```
I feel like the part is a little bit too long. If there is another parser class, it would make the code much easier to understand.

jinhanfromNUS commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,3 @@
+public class DukeException {
```
I think it would be better if we delete this part, as it is empty and it might cause confusion in the future :).

jinhanfromNUS commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,91 @@
+public class Storage {
+
+    static String HOME = System.getProperty("user.home");
+    static Path FILEPATH = Paths.get(HOME, "Downloads", "NUS_CS",
```
I am not sure if I am missing something. This part looks like it is using an absolute path, which might easily create bugs for other users without the same path.

jinhanfromNUS commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,91 @@
+    static Path FILEPATH = Paths.get(HOME, "Downloads", "NUS_CS",
+            "CS2103", "duke.txt");
+
+    public static ArrayList<Task> retrieveList() {
```
Perhaps add a throw exception when file path is not found? 🤔

jinhanfromNUS reviewed (commented):
Keep up the good work!

AwJoey committed:
Add Ui.java: to consolidate System IO functions in a place

AwJoey committed:
Command.java: Enum class for saving command names and description

AwJoey committed:
Add Parser.java: Process user command

AwJoey committed:
Add TaskList.java: Consolidate operations to do on task list

AwJoey committed:
Add Storage.java: Operations to save/retrieve tasks from local file

AwJoey committed:
Add DukeException.java: Custom exception

AwJoey committed:
Edit Deadline.java, Duke.java, Event.java: modifications to cater to new
code consolidation

AwJoey committed:
Organize *.java files into packages

AwJoey committed:
Merge commit 'bdfdc45cfd515daeb10d60e96f5ca72393cd4bc9' into add-gradle-support

AwJoey committed:
Merge branch 'add-gradle-support'

AwJoey committed:
Add ParserTest.java, TaskListTest.java: Add JUnit Tests

AwJoey committed:
Edit build.grade: Package app as JAR file

AwJoey committed:
Add JavaDoc comments to all non-private and non-trivial classes/methods

AwJoey committed:
Revert "Add JavaDoc comments to all non-private and non-trivial classes/methods"

This reverts commit a81d0a1baac53c96ba8a555d4825183f79b78b6e.

AwJoey committed:
Edit *.java: Follow Java coding standard

AwJoey committed:
Edit Command.java, Parser.java, TaskList.java, Ui.java:
Implement find function for users to find tasks by searching for a
keyword

AwJoey committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/task/TaskList.java

AwJoey committed:
Merge branch 'branch-A-JavaDoc'
# Conflicts:
# src/main/java/duke/command/Command.java
# src/main/java/duke/task/Parser.java

AwJoey committed:
Edit Command.java, Parser.java:
Fix bug which runs when Event input is wrong

chuahjiajie commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,90 @@
+import java.time.DateTimeException;
```
Nice parser implementation, will use as inspiration.

chuahjiajie reviewed (commented):
Happy New Year

AwJoey committed:
Add DialogBox.java, Launcher.java: Implement JavaFX

AwJoey committed:
Edit build.gradle

AwJoey committed:
Edit Duke.java

AwJoey committed:
Edit Command.java, Parser.java function

AwJoey committed:
Fix bugs for Level-10 implementation

AwJoey committed:
Revert changes

AwJoey committed:
Merge branch 'branch-Level-10' of https://github.com/AwJoey/ip into branch-Level-10

# Conflicts:
#	src/main/java/duke/parser/Parser.java

AwJoey committed:
Revert back to before testing

AwJoey committed:
Merge branch 'branch-Level-10' of https://github.com/AwJoey/ip into branch-Level-10

AwJoey committed:
Edit Duke.java: Set up Java FX

AwJoey committed:
Merge branch 'branch-Level-10' of https://github.com/AwJoey/ip into branch-Level-10

# Conflicts:
#	src/main/java/duke/Duke.java

AwJoey committed:
Merge changes

AwJoey committed:
Edit Duke: Level-10 Part 2

AwJoey committed:
Edit Duke.java: JavaFX Part 2 Iteration 1 (Echoing user)

AwJoey committed:
Add DaDuke.png, DaUser.png

AwJoey committed:
Edit Duke.java: JavaFX Part 2 Iteration 2 (Dialog Box)

AwJoey committed:
Edit DialogBox.java, Duke.java: JavaFX Part 3 Iteration 3
(DialogBox custom behavior)

AwJoey committed:
Add Command.java classes: Separate methods for easier readability

AwJoey committed:
Edit ParserTest.java: Bug fix

AwJoey committed:
Test new function: ToDo Command implemented

AwJoey committed:
Deadline command implemented in new way

AwJoey committed:
Event command implemented in new way

AwJoey committed:
Implement mark and unmark functions

AwJoey committed:
Implement new find and list function

AwJoey committed:
Implement new Delete and Bye Command

AwJoey committed:
Edit Duke.java: JavaFX Part 4

AwJoey committed:
Polish code

AwJoey committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/task/TaskList.java

AwJoey committed:
Bug fix: merge branch-9

AwJoey committed:
Edit TaskList.java: Add assertions

To prevent bugs. Assertions are used to define assumptions about the
program state so that the runtime can verify them. An assertion failure
indicates a possible bug in the code because the code has resulted in a
program state that violates an assumption about how the code should
behave.

AwJoey committed:
Edit CommandEnum.java, Parser.java, Storage.java, TaskList.java:

Made changes according to comments provided by SE Bot Red

This is to ensure that code will be consistent throughout, especially
when working on tP

AwJoey committed:
Revert "Edit CommandEnum.java, Parser.java, Storage.java, TaskList.java:"

This reverts commit 79d3290846d5299648b8910a34322263356e413a.

AwJoey committed:
Edit CommandEnum.java, Parser.java, Storage.java, TaskList.java:

Made changes according to comments provided by SE Bot Red

This is to ensure that code will be consistent throughout, especially
when working on tP

AwJoey committed:
Merge pull request #2 from AwJoey/branch-A-Assertions

Edit TaskList.java: Add assertions

AwJoey committed:
Merge pull request #3 from AwJoey/branch-A-CodeQuality

Edit CommandEnum.java, Parser.java, Storage.java, TaskList.java:

AwJoey committed:
Add SortCommand.java: Implement sort function to sort tasks by name

AwJoey committed:
Remove Duke in package and classes and edited it to Axolotl

to better reflect the name of the edited bot

AwJoey committed:
Add Ui png and set a threadscheduledexecutor to exit application on bye

AwJoey committed:
Do user guide

future users will know how to interact with Axolotl

AwJoey committed:
Update README.md

AwJoey committed:
Fix bug in the application to polish and add welcome message

ready for v2 release

AwJoey committed:
Edit storage home and filepath variables to make it relative

AwJoey committed:
Edit Main Java class to display title and update build

AwJoey committed:
Handle merge conflict in docs readme file
